src/main/kotlin/Main.kt

```diff
@@ -2,98 +2,192 @@ import com.beust.klaxon.JsonArray
-        var unknown = 0
+    var passed = 0
+    var failed = 0
+    var unknown = 0
 
-        results.forEach { when(it.result.result) {
-            it.result.expected ->  passed++
+    results.forEach {
+        when (it.result.result) {
+            it.result.expected -> passed++
             else -> if (it.result.result == ResultType.EXCEPTION) unknown++ else failed++
-            }
         }
+    }
 
-        println()
-        println("${passed}/$numTests tests succeeded (${(passed*100.0/numTests).roundToInt()}%)" +
+    println()
+    println(
+        "${passed}/$numTests tests succeeded (${(passed * 100.0 / numTests).roundToInt()}%)" +
                 (if (unknown == 0) "" else " ($unknown failed due to exceptions)") +
-                " in ${time/1000} seconds")
+                " in ${time / 1000} seconds"
+    )
+    return results
+}
+
+private fun generateTests(): Collection<Test> {
+    val allRelations = ProofSearcher().addAllProofs().findNewRelations(RelationLoader.relations)
+    //ProofLimiter(3).limit(allRelations)
+    return TestGenerator().addAllTests().generateTests(allRelations)
+}
+
+private fun sortTests(engine: EngineConfiguration, tests: Collection<Test>) : Collection<Test> {
+    val operators = listOf("||", "\\\\", "&&", "consistency:", "refinement:")
+    var out = ArrayList(tests)
 
+    if (engine.queryComplexity != null) { //Query Complexity
+        val (lower, upper) = engine.bounds()
 
-        fullResults.addAll(results)
+        out = ArrayList(out.filter { x ->
+            x.queries().all { y ->
+                y.occurrences(operators) in lower..upper
+            }
+        })
     }
-    return fullResults
+
+    if (engine.testCount != null) {  //Count
+        out = when (engine.testSorting) {
+            Sorting.FILO -> ArrayList(out.takeLast(engine.testCount))
+            Sorting.FIFO -> ArrayList(out.take(engine.testCount))
+            Sorting.RoundRobin -> getEqualTests(out, engine.testCount)
+            Sorting.Random, null -> ArrayList(out.shuffled().take(engine.testCount))
+        }
+    }
+    return out
+}
+
+private fun String.occurrences(strings: Collection<String>): Int {
+    var count = 0
+    for (string in strings) {
+        count += this.occurrences(string)
+    }
+    return count
+}
+
+private fun String.occurrences(string: String): Int
+    = this.windowed(string.length).count { it == string }
+
+private fun getEqualTests(tests: Collection<Test>, count: Int): ArrayList<Test> {
+    val map: HashMap<Pair<String, String>, ArrayList<Test>> = HashMap()
+    tests.forEach { x -> map.getOrPut(Pair(x.type, x.testSuite)) { ArrayList() }.add(x) }
+    println(map.keys)
+    val out = ArrayList<Test>()
+    for (test in map.values.toList()) {
+        out.addAll(test.take(count / map.keys.size))
+    }
+    return out
+}
+
+private fun saveTests(engineName: String, path: String, tests: Collection<Test>) {
+    val file = File(Paths.get(path).absolutePathString()).normalize()
```

Comment:
We need to consider files which cannot be written to, and for this reason i believe we should have this check and throw an exception if we cannot write the file. I think the ``File`` type has the function ``canWrite()``